### ISSUE TITLE
On the path advertisement

### ISSUE THREAD

**Issue by SangameswaranRS** [harnen/service-discovery-paper#12]
1. IIUC, when we walk towards the target, we choose a random node and get a registration. Is it a random node in the bucket or the node in the bucket which is closest to target?

2. It is mentioned that it's the node's responsibility to get the ticket's registered (atleast K of them) - How does the priority work here? I hope we should prioritize closer buckets to the target, but If the waiting time is too high (Diversity + How much the queue is full) we may prioritize the buckets far from target.

I apologize if this isn't the right place to ask such questions, but couldn't find a way to contact the researchers.

**Comment by harnen** [comment]
No worries at all!

Ad. 1 I'm not sure whether you mean the lookup or the registration process. However, for both we want to contact a fixed amount of nodes **per bucket**. This means that for every bucket you'll select a few random nodes. See Algorithm 1 line 6 for registrations and Algorithm 2 line 6 for lookups.

Ad. 2 For the registrations, you want to keep `K_register` active registrations in every bucket, so you don't prioritize. As you correctly state, the registrars close to popular services might get congested and will issue high waiting times. This is fine, you'll get there eventually and your ads will be also present on less congested registrars in further buckets. 

I hope that helps!

**Comment by SangameswaranRS** [comment]
> Ad. 2 For the registrations, you want to keep K_register active registrations in every bucket, so you don't prioritize.

Understood, we keep k active subscription in every bucket. So the skew in concentration will be due to the implicit xor ordering.

If ```Common prefix length``` is considered as a dist metric, bucket 0 will have more peers than bucket 255, as we have the same requirement of maintaining x registrations per bucket, bucket 255 (less peers) will likely have more ads.

> This means that for every bucket you'll select a few random nodes. See Algorithm 1 line 6 for registrations and Algorithm 2 line 6 for lookups.
Could not find any in https://github.com/harnen/service-discovery-paper/blob/main/previous_submissions/eurosp.pdf . Guess I should take the latest and compile them.

**Comment by harnen** [comment]
> But I'm confused as to how this would solve the attacker grinding problem? if the attacker occupies buckets 250 to 255 say, then the probability of a normal node finding providers for the ad would also diminish right? But yes, it won't be zero as we have some ads placed on all buckets.

You're right. However, we start at bucket 0, advance 1 bucket at a time and stop when we find enough ads for our application. As a result, the chance of you even reaching buckets 250-255 is extremely low (the lower, the more popular your application is). An optimal strategy for the attacker is to uniformly spread its Sybil nodes in the network. However, even then, it requires more nodes than all the honest participants has (see Figure 10 of the [eurosp submission](https://github.com/harnen/service-discovery-paper/blob/main/previous_submissions/eurosp.pdf)). 

> Guess I should take the latest and compile them.
We're now working on an improved version of the paper. It's a bit messy at the moment but has some algorithms that (hopefully) make some procedure clearer.

Thanks for the feedback!
